ui/src/shared/apis/queryCache.ts

```diff
@@ -0,0 +1,164 @@
+// Libraries
+import {sortBy} from 'lodash'
+
+// Utils
+import {asAssignment, getAllVariables} from 'src/variables/selectors'
+import {buildVarsOption} from 'src/variables/utils/buildVarsOption'
+import {filterUnusedVarsBasedOnQuery} from 'src/shared/utils/filterUnusedVars'
+import {runQuery} from 'src/shared/apis/query'
+
+// Types
+import {RunQueryResult} from 'src/shared/apis/query'
+import {CancelBox} from 'src/types/promises'
+import {AppState, Variable} from 'src/types'
+
+export const TIME_INVALIDATION = 5000
+
+const asSimplyKeyValueVariables = (vari: Variable) => {
+  return {
+    [vari.name]: vari.selected || [],
+  }
+}
+
+// Hashing function found here:
+// https://jsperf.com/hashcodelordvlad
+// Through this thread:
+// https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript
+const hashCode = (queryText: string): string => {
+  let hash = 0,
+    char
+  if (!queryText) {
+    return `${hash}`
+  }
+  for (let i = 0; i < queryText.length; i++) {
+    char = queryText.charCodeAt(i)
+    hash = (hash << 5) - hash + char
+    hash |= 0 // Convert to 32bit integer
+  }
+  return `${hash}`
+}
+
+class QueryCache {
+  cache = {}
+
+  _cleanExpiredQueries = (): void => {
+    const now = Date.now()
+    for (const id in this.cache) {
+      if (this.cache[id].isCustomTime) {
+        continue
+      }
+      if (now - this.cache[id].dateSet > TIME_INVALIDATION) {
+        this.resetCacheByID(id)
+      }
+    }
+  }
+
+  getFromCache = (id: string, variableID: string): RunQueryResult | null => {
+    // no existing query match
+    if (!this.cache[id]) {
+      return null
+    }
+    // query match with no existing variable match
+    if (this.cache[id].variables !== variableID) {
+      this.resetCacheByID(id)
+      return null
+    }
+    // query & variable match with an expired result
+    if (Date.now() - this.cache[id].dateSet > TIME_INVALIDATION) {
```

Comment:
do you think this expression
`Date.now() - this.cache[id].dateSet > TIME_INVLIADATION`

might be worth abstracting into a named function? `isCacheExpired` or something like that?

Reply:
Since we’re dealing with relative times, I’m not sure how abstracting a method that handles time might affect the accuracy of this condition

Reply:
Great to know. Truthfully, if this condition isn't met at nanosecond precision doesn't necessarily bother me (although @drdelambre might have some thoughts on that) since a false positive will be a definite negative when the watchDog is next run (every 2.5s). You do bring up a good point, and it is something to bear in mind as we roll this out - especially if we find that a 5s caching threshold is too high.